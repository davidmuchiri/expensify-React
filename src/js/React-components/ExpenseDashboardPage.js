import React from 'react';
import ExpenseList from './expenseList';
import ExpenseListFilters from './expenseListFilters';

const ExpenseDashboardPage = () => (
	<div>
		<ExpenseListFilters />
		<ExpenseList />
	</div>
);

export default ExpenseDashboardPage;
