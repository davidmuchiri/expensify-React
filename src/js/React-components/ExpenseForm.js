import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';

const now = moment();
console.log(now.format('Do MMM, YYYY'));

export default class ExpenseForm extends Component {
	constructor(props) {
		super(props);
		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? props.expense.amount.toString() : '',
			createdAt: props.expense
				? moment(props.expense.createdAt)
				: moment(),
			calenderFocused: false,
			error: '',
			buttonText: props.expense ? 'Update Expense' : 'Add Expense'
		};
	}

	onDescriptionChange = e => {
		const description = e.target.value;
		this.setState(() => ({ description }));
	};
	onNoteChange = e => {
		const note = e.target.value;
		this.setState(() => ({ note }));
	};
	onAmountChange = e => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	};
	onDateChange = createdAt => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	};
	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calenderFocused: focused }));
	};

	onSubmit = e => {
		e.preventDefault();
		if (!this.state.description || !this.state.amount) {
			let error = 'Please provide description and amount';
			this.setState(() => ({ error }));
		} else {
			this.setState(() => ({ error: '' }));
			console.log('submitted');
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount).toFixed(2),
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	};

	render() {
		return (
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.onSubmit}>
					<input
						type="text"
						placeholder="Description"
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input
						type="text"
						value={this.state.amount}
						placeholder="Amount"
						onChange={this.onAmountChange}
					/>
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calenderFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>
					<textarea
						placeholder="Add a note for expense (optional)"
						value={this.state.note}
						onChange={this.onNoteChange}
					/>
					<button>{this.state.buttonText}</button>
				</form>
			</div>
		);
	}
}
