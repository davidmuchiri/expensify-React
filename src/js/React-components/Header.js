import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../redux/actions/auth';

export const Header = ({ startLogout }) => (
	<header>
		<h1>Expensify</h1>
		<NavLink to="/dashboard" activeClassName="is-active">
			Dasboard
		</NavLink>
		<NavLink to="/create" activeClassName="is-active">
			create Expense
		</NavLink>
		<button onClick={startLogout}>Logout</button>
	</header>
);

const mapDispatchToProps = dispatch => ({
	startLogout: () => dispatch(startLogout())
});

export default connect(
	undefined,
	mapDispatchToProps
)(Header);
