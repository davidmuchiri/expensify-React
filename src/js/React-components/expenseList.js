import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './expenseListItem';
import selectExpenses from '../redux/selectors/expensesfilter';

export const ExpenseList = props => (
	<div>
		{props.expenses.length === 0 ? (
			<p>No expenses</p>
		) : (
			(<h1> Expense List</h1>,
			props.expenses.map(expense => (
				<ExpenseListItem key={expense.id} {...expense} />
			)))
		)}
	</div>
);

const mapStateToProps = state => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	};
};

export default connect(mapStateToProps)(ExpenseList);
