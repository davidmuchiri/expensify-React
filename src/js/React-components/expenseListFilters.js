import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/initialize';
import {
	setTextFilter,
	sortByDate,
	sortByAmount,
	setStartDate,
	setEndDate
} from '../redux/actions/filters';

const mapStateToProps = state => {
	return {
		filters: state.filters
	};
};
class ExpenseListFilters extends Component {
	state = {
		calendarFocused: null
	};

	onDatesChange = ({ startDate, endDate }) => {
		this.props.dispatch(setStartDate(startDate));
		this.props.dispatch(setEndDate(endDate));
	};
	onFocusChange = calendarFocused => {
		this.setState(() => ({ calendarFocused }));
	};
	render() {
		return (
			<div>
				<input
					type="text"
					value={this.props.filters.text}
					onChange={e => {
						this.props.dispatch(
							setTextFilter(e.target.value)
						);
					}}
				/>
				<select
					value={this.props.filters.sortBy}
					onChange={e => {
						if (e.target.value === 'date') {
							this.props.dispatch(sortByDate());
						} else if (e.target.value === 'amount') {
							this.props.dispatch(sortByAmount());
						}
					}}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>

				<DateRangePicker
					startDate={this.props.filters.startDate}
					endDate={this.props.filters.endDate}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={() => false}
					showClearDates={true}
					startDateId={'start'}
					endDateId={'end'}
				/>
			</div>
		);
	}
}

export default connect(mapStateToProps)(ExpenseListFilters);
