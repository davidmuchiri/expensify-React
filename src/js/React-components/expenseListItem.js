import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, note, amount, createdAt }) => (
	<ul>
		<li>
			<h5>description: {description}</h5>
			{!!note && <p>note: {note}</p>}
			<p>amount: {amount}</p>
			<p>created at: {createdAt}</p>
			<Link to={`/edit/${id}`}>
				<p>Edit</p>
			</Link>
		</li>
	</ul>
);

export default ExpenseListItem;
