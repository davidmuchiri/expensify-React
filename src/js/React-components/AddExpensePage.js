import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../redux/actions/expenses';

export class AddExpensePage extends Component {
	onSubmit = expense => {
		this.props.startAddExpense(expense);
		this.props.history.push('/dashboard');
	};
	render() {
		return (
			<div>
				<h1>Add expense page</h1>
				<ExpenseForm onSubmit={this.onSubmit} />
			</div>
		);
	}
}
const mapDispatchToProps = dispatch => ({
	startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(
	undefined,
	mapDispatchToProps
)(AddExpensePage);
