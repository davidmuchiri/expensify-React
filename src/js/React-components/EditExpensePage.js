import React, { Component } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import ExpenseForm from "./ExpenseForm";
import FaBack from "react-icons/lib/md/arrow-back";
import {
	startEditExpense,
	startRemoveExpense
} from "../redux/actions/expenses";

export class EditExpensePage extends Component {
	onSubmit = expense => {
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push("/dashboard");
	};

	onRemove = () => {
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push("/dashboard");
	};

	render() {
		return (
			<div className="container">
				<div className="addExpensepage__panel">
					<Link
						to="/dashboard"
						className="addExpensepage__panel--link">
						<FaBack className="icon" />
					</Link>
					<h5 className="addExpensepage__title">
						Edit expense page
					</h5>
				</div>
				<ExpenseForm
					expense={this.props.expense}
					onSubmit={this.onSubmit}
					onRemove={this.onRemove}
				/>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, expense) =>
		dispatch(startEditExpense(id, expense)),
	startRemoveExpense: data => dispatch(startRemoveExpense(data))
});

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find(
		expense => expense.id === props.match.params.id
	)
});

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(EditExpensePage);
