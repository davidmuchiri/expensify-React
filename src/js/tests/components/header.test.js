import React, { Component } from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../React-components/Header';

test('should render Header correctly', () => {
	const wrapper = shallow(<Header startLogOut={() => {}} />);
	expect(wrapper).toMatchSnapshot();
});

//should call startLogout on button click
test('should call startLogout on button click', () => {
	const startLogout = jest.fn();
	const wrapper = shallow(<Header startLogout={startLogout} />);
	wrapper.find('button').simulate('click');
	expect(startLogout).toHaveBeenCalled();
});

//loginpage test files => should call startLogin on button clicks
