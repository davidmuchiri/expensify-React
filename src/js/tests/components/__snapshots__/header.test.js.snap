// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    to="/dashboard"
  >
    Dasboard
  </NavLink>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    to="/create"
  >
    create Expense
  </NavLink>
  <button>
    Logout
  </button>
</header>
`;
